pkg/jobmanager/job.go

```diff
@@ -0,0 +1,258 @@
+	programName   string
+	programArgs   []string
+	cmd           *exec.Cmd
+	stdoutBuffer  io.OutputBuffer
+	stderrBuffer  io.OutputBuffer
+	running       bool
+	runErrors     []error
+}
+
+// NewJob creates and returns a new concreteJob based on the given values.
+func NewJob(
+	owner string,
+	name string,
+	cgControllers []cgroup.Controller,
+	programName string,
+	programArgs ...string,
+) Job {
+
+	return NewJobDetailed(
+		owner,
+		name,
+		cgControllers,
+		io.NewMemoryBuffer(),
+		io.NewMemoryBuffer(),
+		programName,
+		programArgs...,
+	)
+}
+
+func NewJobDetailed(
+	owner string,
+	name string,
+	cgControllers []cgroup.Controller,
+	stdoutBuffer io.OutputBuffer,
+	stderrBuffer io.OutputBuffer,
+	programName string,
+	programArgs ...string,
+) Job {
+
+	return &concreteJob{
+		owner:         owner,
+		id:            uuid.New(),
+		name:          name,
+		cgControllers: cgControllers,
+		programName:   programName,
+		programArgs:   programArgs,
+		stdoutBuffer:  stdoutBuffer,
+		stderrBuffer:  stderrBuffer,
+	}
+}
+
+// Start starts the job if it hasn't already been started.
+func (j *concreteJob) Start() error {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+
+	if j.cmd != nil {
+		return fmt.Errorf("job %s (%v) has already been started", j.name, j.id)
+	}
+
+	cgroupSet := cgroup.NewSet(j.id, j.cgControllers...)
+	if err := cgroupSet.Create(); err != nil {
+		return err
+	}
+
+	args := cgroupSet.TaskFiles()
+	args = append(args, "--")
+	args = append(args, j.programName)
+	args = append(args, j.programArgs...)
+
+	j.cmd = exec.Command(config.CgexecPath, args...)
+	j.cmd.Stdout = j.stdoutBuffer
+	j.cmd.Stderr = j.stderrBuffer
+	j.cmd.Env = make([]string, 0) // Do not pass along our environment
+
+	//j.cmd.Dir = "/" // If we were to chroot
+
+	j.cmd.SysProcAttr = &syscall.SysProcAttr{
+		Chroot: "", // This would be non-empty to actually do a chroot
+		Cloneflags: syscall.CLONE_NEWPID |
+			syscall.CLONE_NEWNS |
+			syscall.CLONE_NEWNET,
+	}
+
+	j.running = true
+
+	go func() {
+		defer j.lockedOperation(func() {
+			if err := cgroupSet.Destroy(); err != nil {
+				j.runErrors = append(j.runErrors, err)
+			}
+			j.running = false
+		})
+
+		// Run blocks until the newly-created process terminates.  It calls
+		// Wait internally
+		err := j.cmd.Run()
+
+		// Once Wait returns, all output has been written to Stdout and Stderr
+		j.lockedOperation(func() {
+			if err != nil {
+				j.runErrors = append(j.runErrors, err)
+			}
+
+			if err := j.stdoutBuffer.Close(); err != nil {
+				j.runErrors = append(j.runErrors, err)
+			}
+
+			if err := j.stderrBuffer.Close(); err != nil {
+				j.runErrors = append(j.runErrors, err)
+			}
+		})
+	}()
+
+	return nil
+}
+
+// Stop kills the job.
+func (j *concreteJob) Stop() error {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+
+	if !j.running {
+		// If the job isn't running, it is stopped already
+		return nil
+	}
+
+	if err := j.cmd.Process.Kill(); err != nil && err != os.ErrProcessDone {
```

Comment:
Will this kill all spawned processes?

Reply:
Had we not run the process in a new pid namespace, it would not.  However, the child process here is the init process within a newly-created pid namespace.  If that init process terminates, the kernel will automatically terminate any other processes within the pid namespace.